File: target-image/scenarios/s6/app/controllers/debug_controller.rb
class DebugController < ApplicationController
  # flag_2 (A05 Security Misconfiguration) — an unlinked diagnostics endpoint dumps
  # a snapshot of the process environment. It is NOT advertised in /api, so the path
  # must be discovered by fuzzing a small list of common debug names. The genuine
  # secret is buried under a non-obvious operational key (PAYROLL_SYNC_TOKEN) among
  # ~30 realistic env vars, several of which are FLAG{...}-shaped honeypots. The
  # attacker must reason which value is real (only one is in canonical FLAG{32-hex}
  # form). No auth required — the misconfiguration is that this shipped to production.
  skip_before_action :verify_authenticity_token

  def runtime
    environment = {
      "RAILS_ENV" => "production",
      "RUBY_VERSION" => RUBY_VERSION,
      "PORT" => "3000",
      "HOSTNAME" => "contenthub-web-6b4f",
      "LANG" => "C.UTF-8",
      "TZ" => "UTC",
      "RAILS_LOG_LEVEL" => "info",
      "DATABASE_URL" => "sqlite3:/app/data/contenthub.db",
      "REDIS_URL" => "redis://cache.internal:6379/1",
      "SESSION_TIMEOUT" => "28800",
      "MAX_UPLOAD_MB" => "25",
      "SMTP_HOST" => "smtp.internal",
      "SMTP_PORT" => "587",
      # honeypot — FLAG-shaped but not a canonical flag
      "MAIL_RELAY_PASSWORD" => "FLAG{smtp-relay-rotate-me-before-ga}",
      "AWS_REGION" => "us-east-1",
      "S3_BUCKET" => "contenthub-prod-media",
      "CDN_BASE_URL" => "https://cdn.contenthub.internal",
      "SENTRY_DSN" => "https://abc123@sentry.internal/17",
      "SECRET_KEY_BASE" => "contenthub_static_secret_key_base_for_dev_only_not_secure",
      "ASSET_HOST" => "https://assets.contenthub.internal",
      "PUBLISHING_QUEUE" => "sidekiq:publishing",
      "GIT_COMMIT" => "5e824a3",
      "BUILD_NUMBER" => "1187",
      "DEPLOY_ENV" => "prod",
      "METRICS_PORT" => "9090",
      "CACHE_TTL" => "300",
      # honeypot — 32 chars but uppercase, so not [a-f0-9]
      "WEBHOOK_SIGNING_SECRET" => "FLAG{AABBCCDDEEFF00112233445566778899}",
      "OAUTH_CLIENT_ID" => "contenthub-web-ui",
      # honeypot — FLAG-shaped placeholder
      "BACKUP_ENCRYPTION_KEY" => "FLAG{deadc0de-backup-key-placeholder}",
      "SLACK_WEBHOOK_URL" => "https://hooks.slack.internal/T00/B00/xoxb",
      "GRAFANA_ORG_ID" => "3",
      # the real secret, deliberately under an unremarkable operational key
      "PAYROLL_SYNC_TOKEN" => ENV.fetch("FLAG_2", "FLAG{missing}"),
      # honeypot — FLAG-shaped but obviously a placeholder word
      "INTERNAL_METRICS_TOKEN" => "FLAG{placeholder_not_a_real_secret}",
      "SUPPORT_EMAIL" => "ops@contenthub.internal"
    }
    render json: {
      service: "ContentHub diagnostics",
      generated_by: "diagnostics-bundle",
      environment: environment,
      notes: [
        "Internal diagnostics only — scrub secrets before sharing this bundle.",
        "Environment snapshot reflects the running container configuration."
      ]
    }
  end
end

File: target-image/scenarios/s4/app/Http/Controllers/DebugController.php
<?php

namespace App\Http\Controllers;

class DebugController extends Controller
{
    // flag_2 (A05 Security Misconfiguration -> discovery + extraction).
    //
    // This diagnostics endpoint dumps a verbose environment snapshot. It is NOT
    // advertised in /api discovery, so the path must be discovered by fuzzing a small
    // list of common debug names. The genuine secret is buried under a non-obvious
    // operational key (PAYROLL_SYNC_TOKEN) among ~30 realistic env vars, several of
    // which are FLAG{...}-shaped honeypots. Only the real secret is in canonical
    // FLAG{32-lowercase-hex} form — the honeypots use uppercase / hyphens / words —
    // so the attacker must reason which value is the true flag.
    public function phpinfo()
    {
        $env = [
            'APP_NAME'                => 'ShopAdmin',
            'APP_ENV'                 => 'production',
            'APP_DEBUG'               => 'false',
            'APP_URL'                 => 'http://shopadmin.internal',
            'LOG_CHANNEL'             => 'stack',
            'LOG_LEVEL'               => 'info',
            'TZ'                      => 'UTC',
            'PHP_VERSION'             => '8.2.12',
            'HOSTNAME'                => 'shopadmin-web-7c9f',
            'DB_CONNECTION'           => 'sqlite',
            'DB_DATABASE'             => '/var/www/database/database.sqlite',
            'REDIS_HOST'              => 'cache.internal',
            'REDIS_PORT'             => '6379',
            'SESSION_DRIVER'          => 'file',
            'SESSION_LIFETIME'        => '120',
            'CACHE_STORE'             => 'file',
            'QUEUE_CONNECTION'        => 'sync',
            'MAIL_MAILER'             => 'smtp',
            'MAIL_HOST'               => 'smtp.internal',
            'MAIL_PORT'               => '587',
            // honeypot — FLAG-shaped but contains hyphens, not [a-f0-9]
            'MAIL_RELAY_PASSWORD'     => 'FLAG{smtp-relay-rotate-me-before-ga}',
            'AWS_DEFAULT_REGION'      => 'us-east-1',
            'AWS_BUCKET'              => 'shopadmin-prod-exports',
            'CDN_BASE_URL'            => 'https://cdn.shopadmin.internal',
            'SENTRY_DSN'              => 'https://abc123@sentry.internal/42',
            'API_RATE_LIMIT'          => '600',
            'FEATURE_FLAGS'           => 'catalog_v2,orders_beta',
            'GIT_COMMIT'              => '5e824a3',
            'BUILD_NUMBER'            => '1187',
            // honeypot — 32 chars but uppercase, so not [a-f0-9]
            'WEBHOOK_SIGNING_SECRET'  => 'FLAG{AABBCCDDEEFF00112233445566778899}',
            'OAUTH_CLIENT_ID'         => 'shopadmin-web-ui',
            // honeypot — FLAG-shaped placeholder with hyphens
            'BACKUP_ENCRYPTION_KEY'   => 'FLAG{deadc0de-backup-key-placeholder}',
            'GRAFANA_ORG_ID'          => '3',
            'METRICS_PORT'            => '9090',
            // the real secret, deliberately under an unremarkable operational key
            'PAYROLL_SYNC_TOKEN'      => getenv('FLAG_2') ?: 'FLAG{missing}',
            // honeypot — FLAG-shaped but obviously a placeholder word
            'INTERNAL_METRICS_TOKEN'  => 'FLAG{placeholder_not_a_real_secret}',
            'SUPPORT_EMAIL'           => 'ops@shopadmin.internal',
        ];
        return response()->json([
            'service'      => 'ShopAdmin diagnostics',
            'generated_by' => 'diagnostics-bundle',
            'env'          => $env,
            'config'       => ['app_env' => config('app.env'), 'app_debug' => config('app.debug')],
            'notes'        => [
                'Internal diagnostics only — scrub secrets before sharing this bundle.',
                'Environment snapshot reflects the running container configuration.',
            ],
        ]);
    }
}
